Rebus.AzureTables/AzureTables/Subscriptions/AzureTablesSubscriptionStorage.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Data.Tables;
+using Rebus.Bus;
+using Rebus.Exceptions;
+using Rebus.Subscriptions;
+
+namespace Rebus.AzureTables.Subscriptions;
+/// <summary>
+/// Implementation of <see cref="ISubscriptionStorage"/> that uses Azure tables (either Azure Cosmos Db Table Api or Azure Table Storage) to store subscriptions
+/// </summary>
+public class AzureTablesSubscriptionStorage : ISubscriptionStorage, IInitializable {
+  private readonly TableClient _tableClient;
+  private readonly bool _automaticallyCreateTable;
+  private bool _tableInitialized;
+
+  /// <summary>
+  /// Creates the subscription storage
+  /// </summary>
+  public AzureTablesSubscriptionStorage(TableClient tableClient, bool isCentralized = false, bool automaticallyCreateTable = false) {
+    _tableClient = tableClient;
+    IsCentralized = isCentralized;
+    _automaticallyCreateTable = automaticallyCreateTable;
+  }
+
+  /// <summary>
+  /// Initializes the subscription storage by ensuring that the necessary table is created
+  /// </summary>
+  public void Initialize() => GetTableClient().Wait();
+
+  /// <summary>
+  /// Gets all subscribers by getting row IDs from the partition named after the given <paramref name="topic"/>
+  /// </summary>
+  public async Task<string[]> GetSubscriberAddresses(string topic) {
+    try {
+      var client = await GetTableClient();
+      var items = client.QueryAsync<TableEntity>(s => s.PartitionKey == topic, select: new[] { "PartitionKey", "RowKey" });
+      var addresses = new List<string>();
+      await foreach (var address in items)
+        addresses.Add(address.RowKey);
+      return addresses.ToArray();
```

Comment:
In C# you should never return `Task<IEnumerable<T>>`, because it insinuatis that it's a double-future (i.e. lazy-materialized on two levels), which is kind of wonky. In this case, the list IS in fact materialized, so `Task<IReadOnlyList<string>>` would probably be the best return type, because then there would be no further copying of arrays involved.

Reply:
Interesting thought, although I'm not sure that `IEnumerable` necessarily signals that the returned value will be a lazy-evaluated one. Perhaps you have an article to share that tries explaining this? Just curious.

Also - you already have a `List<string>`, why do you convert it into an array (`.ToArray()`)?